python/pyarrow/_flight.pyx

```diff
@@ -1597,6 +1644,104 @@ cdef CStatus _get_token(void* self, c_string* token) except *:
     return CStatus_OK()
 
 
+cdef CStatus _middleware_sending_headers(
+        void* self, CAddCallHeaders* add_headers) except *:
+    """Callback for implementing middleware."""
+    try:
+        headers = (<object> self).sending_headers()
+    except FlightError as flight_error:
+        return (<FlightError> flight_error).to_status()
+
+    if headers:
+        for header, values in headers.items():
+            if isinstance(values, (six.text_type, six.binary_type)):
```

Comment:
We should decide whether we want to accept text or binary headers. I'd say only binary.

Reply:
However, gRPC would disagree with me:
https://grpc.github.io/grpc/python/glossary.html?highlight=header

Reply:
Only-text is more consistent with REST & HTTP/1.1, and I assumed it would be easier to support for any potential future backends for Flight. (It also avoids the -bin suffixing.)

Reply:
I've changed the Python side to require only ASCII.